pop_tools/fill.py

```diff
@@ -183,3 +223,93 @@ def _iterative_fill_POP_core(nlat, nlon, var, fillmask, missing_value, tol, ltri
+        if zoncnt[j] != 0:
+            zonavg[j] = zonavg[j] / zoncnt[j]
+
+    # Fill missing zonal averages for rows that are entirely land
+    for j in range(0, nlat - 1):  # northward pass
+        if zoncnt[j] > 0 and zoncnt[j + 1] == 0:
+            zoncnt[j + 1] = 1
+            zonavg[j + 1] = zonavg[j]
+    for j in range(0, nlat - 1):  # southward pass
+        jrev = nlat - 1 - j
+        if zoncnt[jrev] > 0 and zoncnt[jrev - 1] == 0:
+            zoncnt[jrev - 1] = 1
+            zonavg[jrev - 1] = zonavg[jrev]
+
+    # Replace the input array missing values with zonal average as first guess
+    for j in range(0, nlat):
+        for i in range(0, nlon):
+            if fillmask[j, i]:
+                var[j, i] = zonavg[j]
+
+    # Now do the iterative 2D fill
+    res = np.zeros((nlat, nlon))  # work array hold residuals
+    res_max = tol
+    iter = 0
+    while iter < max_iter and res_max >= tol:
+        res = res * 0.0  # reset the residual to zero for this iteration
+
+        # assume bottom row is all land, leave it set to zonal average
+        # deal with top row separately below
+        for j in range(1, nlat - 1):
+            jm1 = j - 1
+            jp1 = j + 1
+
+            for i in range(0, nlon):
+                if fillmask[j, i]:
+                    im1 = i - 1
+                    if i == 0:  # assume periodic in x
+                        im1 = nlon - 1
+                    ip1 = i + 1
+                    if i == nlon - 1:
+                        ip1 = 0
+
+                    # this is SOR
+                    res[j, i] = (
+                        var[j, ip1] + var[j, im1] + var[jm1, i] + var[jp1, i] - 4.0 * var[j, i]
+                    )
+                    var[j, i] = var[j, i] + rc * 0.25 * res[j, i]
+
+        # do the top row if there was some valid data there in the input
+        # otherwise leave it set to zonal average of northernmost row with valid data
+        if zoncnt[nlat - 1] > 1:
+            j = nlat - 1
+            jm1 = j - 1
+            jp1 = j
+            for i in range(0, nlon):
+                if fillmask[j, i]:
+                    im1 = i - 1
+                    if i == 0:
+                        im1 = nlon - 1
+                    ip1 = i + 1
+                    if i == nlon - 1:
+                        ip1 = 0
+                    io = nlon - 1 - i
+
+                    if ltripole:  # use cross-pole periodicity
+                        res[j, i] = (
+                            var[j, ip1] + var[j, im1] + var[jp1, io] + var[jm1, i] - 4.0 * var[j, i]
+                        )
+                        var[j, i] = var[j, i] + rc * 0.25 * res[j, i]
+                    else:  # do a 1D smooth on pole row
+                        res[j, i] = var[j, ip1] + var[j, im1] - 2.0 * var[j, i]
+                        var[j, i] = var[j, i] + rc * 0.5 * res[j, i]
+
+        res_max = np.max(np.fabs(res)) / np.max(np.fabs(var))
```

Comment:
Consider replacing
```
        res_max = np.max(np.fabs(res)) / np.max(np.fabs(var))
```
with
```
        res_max = max(abs(res)) / max(abs(var))
```
If you do this, then you could also consider replacing other occurrences of `np.fabs` and `np.abs` with `abs`.

Reply:
> `res_max = max(abs(res)) / max(abs(var))`

Won't work unless `res` is a 1D array:


```python
In [19]: res = np.arange(10)

In [20]: max(abs(res))
Out[20]: 9

In [21]: res = res.reshape(5, 2)

In [22]: max(abs(res))
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-22-597f49d61954> in <module>
----> 1 max(abs(res))

ValueError: The truth value of an array with more than one element is ambiguous. Use a.any() or a.all()
```

Reply:
Thanks for correcting me @andersy005. I had tested the idea, but only on 1D arrays.